File: lib/sys/linux/mknod.ex
defmodule Sys.Linux.Mknod do
  @moduledoc "Stage block-device nodes and files into a VM chroot (via the setuid helper)."

  use OpenTelemetryDecorator

  alias Hyper.SuidHelper

  @doc """
  Create in the chroot a block-device node at `dest` mirroring `device` (a host
  block-device path: `/dev/loopN` or `/dev/mapper/hyper-*`), owned `uid:gid`. The
  helper validates `device` and reads its `major:minor` from the device itself,
  so the caller can never name an arbitrary device.
  """
  @spec block(Path.t(), Path.t(), non_neg_integer(), non_neg_integer()) ::
          :ok | {:error, {non_neg_integer(), String.t()}}
  @decorate with_span("Sys.Linux.Mknod.block", include: [:dest, :device])
  def block(dest, device, uid, gid) do
    # mknod has no underlying binary; SuidHelper.run/3 omits --bin for it.
    case SuidHelper.run("mknod", Hyper.Config.suid_helper(), [
           "--dest",
           dest,
           "--device",
           device,
           "--uid",
           to_string(uid),
           "--gid",
           to_string(gid)
         ]) do
      {:ok, _} -> :ok
      {:error, _} = err -> err
    end
  end

  @doc "Hardlink-or-copy `src` to `dest` inside a chroot, owned `uid:gid`."
  @spec stage(Path.t(), Path.t(), non_neg_integer(), non_neg_integer()) ::
          :ok | {:error, {non_neg_integer(), String.t()}}
  @decorate with_span("Sys.Linux.Mknod.stage", include: [:src, :dest])
  def stage(src, dest, uid, gid) do
    case SuidHelper.run("stage", Hyper.Config.suid_helper(), [
           "--src",
           src,
           "--dest",
           dest,
           "--uid",
           to_string(uid),
           "--gid",
           to_string(gid)
         ]) do
      {:ok, _} -> :ok
      {:error, _} = err -> err
    end
  end
end

File: native/suidhelper/src/tools/mknod.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! `mknod` helper: create a block device node inside a VM chroot.
//!
//! Security model
//! ──────────────
//! 1. `device` is a [`BlockDev`] (lexically restricted to `/dev/loopN` or
//!    `/dev/mapper/hyper-*`): the caller names one of our own devices.
//! 2. We open it as a [`SafeFile<IsBlockDevice, …>`] (`O_PATH|O_NOFOLLOW`): the
//!    type proves, via `fstat`, that it really is a block device, and `rdev()` is
//!    only callable on that proven handle - so the major:minor come from a
//!    verified device node, never a caller-supplied number.
//! 3. `parent` is an already-confined chroot directory fd; `mknod_block` creates
//!    the node and `chown`s it (`AT_SYMLINK_NOFOLLOW`) relative to that fd.

use crate::safe_dev::BlockDev;
use crate::util::safe_dir::{self, SafeDir};
use crate::util::safe_file::{self, Any, IsBlockDevice, SafeFile};
use crate::util::safe_path::{self, IsAbsolute, SafePath, StrictComponents};
use nix::fcntl::OFlag;
use thiserror::Error as ThisError;

#[derive(Debug, ThisError)]
pub enum Error {
    #[error("device path: {0}")]
    DevicePath(#[from] safe_path::ValidationError),
    #[error("device: {0}")]
    Device(#[from] safe_file::ValidationError),
    #[error("mknod {name:?}: {source}")]
    Node {
        name: String,
        #[source]
        source: safe_dir::Error,
    },
}

/// Create a block-device node `name` under `parent`, mirroring `device`'s
/// major:minor, owned `uid:gid`.
pub(crate) fn make_block_node(
    parent: &SafeDir,
    name: &str,
    device: &BlockDev,
    uid: u32,
    gid: u32,
) -> Result<(), Error> {
    let dev_path: SafePath<IsAbsolute, StrictComponents> = device.as_ref().to_path_buf().try_into()?;
    let dev = SafeFile::<IsBlockDevice, Any, Any>::open(&dev_path, OFlag::O_PATH)?;
    let rdev = dev.rdev()?;

    parent
        .mknod_block(name, rdev, uid, gid)
        .map_err(|source| Error::Node { name: name.to_string(), source })
}
